pennylane_lightning/src/algorithms/JacobianProd.hpp

```diff
@@ -122,48 +93,47 @@ class VectorJacobianProduct : public AdjointJacobian<T> {
      * @brief Calculates the VectorJacobianProduct for the statevector
      * for the selected set of parametric gates using `AdjointJacobian`.
      *
-     * @param vjp Preallocated vector for vector-jacobian product data results
-     * of size `trainableParams.size()`.
-     * @param jac Preallocated Jacobian matrix from `AdjointJacobian` of size
-     * `observables.size() * trainableParams.size()`.
-     * @param psi Pointer to the statevector data.
-     * @param num_elements Length of the statevector data.
      * @param dy Gradient-output vector.
-     * @param observables Observables for which to calculate Jacobian.
-     * @param operations Operations used to create given state.
-     * @param trainableParams List of parameters participating in Jacobian
-     * calculation.
-     * @param apply_operations Indicate whether to apply operations to psi prior
-     * to calculation.
+     * @param num_params Total number of parameters in the QuantumTape
+     * @param apply_operations Indicate whether to apply operations to jd.psi
+     * prior to calculation.
+     *
+     * @return std::function<std::vector<T>(const JacobianData<T> &jd)>
+     * where `jd` is a JacobianData object representing the QuantumTape
+     * to differentiate.
+     *
      */
-    void vectorJacobianProduct(std::vector<T> &vjp,
-                               std::vector<std::vector<T>> &jac,
-                               const std::vector<T> &dy,
-                               const std::complex<T> *psi, size_t num_elements,
-                               const std::vector<ObsDatum<T>> &observables,
-                               const OpsData<T> &operations,
-                               const std::vector<size_t> &trainableParams,
-                               bool apply_operations = false) {
-        const size_t num_params = trainableParams.size();
-
-        if (num_params == 0U || dy.empty()) {
-            vjp.clear();
-            return;
+    auto vectorJacobianProduct(const std::vector<T> &dy, size_t num_params,
+                               bool apply_operations = false)
+        -> std::function<std::vector<T>(const JacobianData<T> &)> {
+        if (dy.empty() ||
+            std::all_of(dy.cbegin(), dy.cend(), [](T e) { return e == 0; })) {
+            // If the dy vector is zero, then the
+            // corresponding element of the VJP will be zero,
+            // and we can avoid unnecessary computation.
+            return
+                [&num_params]([[maybe_unused]] const JacobianData<T> &jd)
```

Comment:
Hi Ali, I also think this function returns a lambda with `num_params` (which is local variable) captured as a reference. As it is just `size_t`, I think you can change the capture to` [=]` or `[num_params = num_params]`.

Reply:
Good point! ... DON